packages/fxa-settings/src/lib/context/implementations/storage-context.ts

```diff
@@ -0,0 +1,91 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Context } from "..";
+
+export class StorageContext implements Context {
+
+  private static readonly NAMESPACE = '__fxa_session';
+  private static readonly PERSIST_TO_LOCAL_STORAGE = ['oauth'];
+
+  private state: Record<string, any>;
+
+  constructor(
+    private window: Window
+  ) {
+    this.state = {};
+    this.load();
+  }
+
+  public requiresSync() {
+    return true;
+  }
+
+  public load() {
+    let values = {};
+
+    // Try parsing sessionStorage values
+    try {
+      const sessionStorageValueRaw = this.window.sessionStorage.getItem(StorageContext.NAMESPACE);
+      if (sessionStorageValueRaw != null) {
+        values = {
+          ...values,
+          ...JSON.parse(sessionStorageValueRaw)
+        };
+      }
+    } catch (e) {
+      // ignore the parse error.
+    }
+
+    // Try parsing localStorage values
+    try {
+      const localStorageValueRaw = this.window.localStorage.getItem(StorageContext.NAMESPACE);
+      if (localStorageValueRaw != null) {
+        values = {
+          ...values,
+          ...JSON.parse(localStorageValueRaw)
+        };
+      }
+    } catch (e) {
+      // ignore the parse error.
+    }
+
+    this.state = values;
+  }
+
+  public persist() {
+
+    const toSaveToSessionStorage: Record<string, any> = {};
+    const toSaveToLocalStorage: Record<string, any> = {};
+
+    for (const key in this.state) {
+      const value = this.state[key];
+
+      if (StorageContext.PERSIST_TO_LOCAL_STORAGE.indexOf(key) >= 0) {
+        toSaveToLocalStorage[key] = value;
+      } else {
+        toSaveToSessionStorage[key] = value;
+      }
+    }
+
+    // Wrap browser storage access in a try/catch block because some browsers
+    // (Firefox, Chrome) except when trying to access browser storage and
+    // cookies are disabled.
+    try {
+      this.window.localStorage.setItem(StorageContext.NAMESPACE, JSON.stringify(toSaveToLocalStorage));
+      this.window.sessionStorage.setItem(StorageContext.NAMESPACE, JSON.stringify(toSaveToSessionStorage));
+    } catch (e) {
+      // some browsers disable access to browser storage
+      // if cookies are disabled.
```

Comment:
@LZoog I think we need a follow up for the storage context thing. I'd like to use the storage class that exists in fxa-settings since it seems quite a bit better, but I think that will require migrating the existing models... or modifying how the namespace thing works. This was mentioned in some other places in the review as well, but I'll bring it up again here. Do you want me to file a ticket for this?

Reply:
Follow up created: FXA-6801